[kiwibang] iP

kiwibang opened the pull request:
# Nyanbot
> nyan nyan nyan nyan nyan nyan nyan nyan
> nyan nyan nyan nyan nyan nyan nyan nyan
> -_nyan cat, 2023_

Nyanbot helps you to track your tasks in a nyantastic way. Its
- Text-based
- Lightweight
- Easy to use
- Is _actually_ a cat

All you need to do is:
1. download it from [here](https://github.com/kiwibang/ip/releases/tag/A-Jar)
2. Double-click it
3. Add the tasks that are on your mind
4. Mark tasks as done as your day goes by

And it is **FREE**

Features:
- [x] Todo tasks without date
- [x] Deadline tasks with a due date
- [x] Event tasks with a start date and time
- [ ] A ~~fantastic~~ nyantastic GUI for users that prefer a more visual interface :eyes:

Sample commands:
```
todo decide bread
deadline eat bread /by 25th Dec
event make bread /from 12/12/2023 /to 12/12/2024
```

Here is our `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

kiwibang committed:
Add Level-0 features

kiwibang committed:
Add Level-1 features

kiwibang committed:
Add Level-2 features

kiwibang committed:
Add Level-3 features

kiwibang committed:
Add Level-4 features

kiwibang committed:
Add A-TextUiTesting features

kiwibang committed:
Add Level-5 features

kiwibang committed:
Add Level-6 features

kiwibang committed:
Refactor code 1

Commit:
Add Level-7 features

Commit:
Merge branch 'branch-Level-7'

Commit:
Add Level-8 features

Commit:
Merge branch 'branch-Level-8'

Commit:
Add A-MoreOOP features

Commit:
Merge branch 'branch-A-MoreOOP'

Commit:
Add A-Packages features

Commit:
Merge branch 'add-gradle-support' of ssh://github.com/kiwibang/ip

Commit:
Add A-JUnit features

Commit:
Add A-Jar features

Commit:
Revert "Add A-Jar features"

This reverts commit cb5339014b8f2dc8ce093b798ca218a18d598a62.

Commit:
Add A-CodingStandard features

Commit:
Add A-JavaDoc features

Commit:
Add Level-9 features

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/DukeParsers/DukeParser.java

Commit:
Add Level-10 features

Commit:
Add Gradle setup

Commit:
Merge branch 'branch-Level-10'

Commit:
Add A-CheckStyle features

aslam341 commented on src/main/java/com/nyanbot/DukeGuiControllers/DukeMainWindow.java:
```diff
@@ -0,0 +1,73 @@
+package com.nyanbot.DukeGuiControllers;
```
Should this be in all lower case, as per [coding standard](https://se-education.org/guides/conventions/java/intermediate.html) for naming convention for packages? I noticed the same in several other places too.

aslam341 commented on src/main/java/com/nyanbot/DukeTasks/Task.java:
```diff
@@ -0,0 +1,316 @@
+    // flag to indicate whether tha task has been completed
+    private boolean isDone;
+    // flag to indicate whether the task is valid
+    private boolean isValid;
```
Well done in naming boolean variables to sound like booleans.

aslam341 commented on src/main/java/com/nyanbot/DukeTaskList/DukeTaskList.java:
```diff
@@ -0,0 +1,395 @@
+            } else {
+                int fromStart = task.lastIndexOf("/from");
+                int toStart = task.lastIndexOf("/to");
+                createdTask = this.errorUI.handleEverythingElseError(task, fromStart, toStart, taskDescriptionLength);
```
I like how you placed a line break here to improve readability.

aslam341 commented on src/main/java/com/nyanbot/DukeStorage/DukeStorageDatabase.java:
```diff
@@ -0,0 +1,131 @@
+public class DukeStorageDatabase {
+    // hardcoded file paths
+    private static final String DIR_PATH = System.getProperty("user.dir") + "/data";
+    private static final String FILE_PATH = System.getProperty("user.dir") + "/data/duke.txt";
```
Should the 2 constants have a common prefix, as opposed to a common suffix? This would be in accordance to the [coding standard](https://se-education.org/guides/conventions/java/intermediate.html) where it states that "Associated constants should have a common prefix".

aslam341 commented on src/test/Java/DukeGuiControllers/DukeTest.java:
```diff
@@ -0,0 +1,93 @@
+    // inputStream for the system.in
+    private static final InputStream systemIn = System.in;
+    // outputStream for System.out
+    private static final PrintStream systemOut = System.out;
```
Should these constant names be all uppercase, using underscore to separate words, as per [coding standard](https://se-education.org/guides/conventions/java/intermediate.html)?

aslam341 reviewed (commented):
LGTM. Just some potential nits to fix.

XihuaZ commented on src/test/Java/DukeGuiControllers/DukeTest.java:
```diff
@@ -0,0 +1,93 @@
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import java.io.*;
```
What specific classes is it supposed to import?

XihuaZ commented on src/main/java/com/nyanbot/DukeTasks/Task.java:
```diff
@@ -0,0 +1,316 @@
+     *         set to 23 and minute field set to 59
+     */
+    private LocalDateTime addTime(LocalDateTime time) {
+        return time.withHour(23).withMinute(59);
```
For the integers such as 23 and 59, maybe it would be better to create a private static final class for them? since what they represent is not apparent at first glance.

XihuaZ reviewed (commented):
Besides the issue of using wildcards, it is a well written piece of code. Keep up the good work!

Commit:
Refactor package names to lowercase

Commit:
Reinstate JUnit to project

Commit:
Reinstate checkstyle and fix java coding standard violations

Commit:
Generate new .jar file for week 5

Commit:
Generate new .jar file for week 5

This commit generates a jar file for the current project after adding
the relevant fixes.

The jar file can be run as a separate application.

Commit:
merge commit message branch back to master

This merges the commit message branch to master.

I am not sure why SourceTree created a new branch,
but I am fixing it through a merge commit.

Commit:
Add A-Assertions features

kiwibang committed:
Merge pull request #1 from kiwibang/branch-A-Assertions

Add A-Assertions features

Commit:
Add A-CodeQuality features

This branch fixes some code quality issues.

It is highly likely that I missed out some code quality
issues, since there are too many to keep track of.

kiwibang committed:
Merge pull request #2 from kiwibang/branch-A-CodeQuality

Add A-CodeQuality features

Commit:
Add A-Streams features

Commit:
Refactor time parser utilities

This commit refactors away the time parser utilities that were present
inItially in the Task class.

Why I did so was because it did not make sense for a Task class to contain
date parsing utilities.

Commit:
Add BCD-Extension features

I chose C-NaturalDates since I have already implemented
it as part of Level-8. Hence, my iP already has this extension.

Hence, as a programmer that wants to save
my most important resource(i.e. time),
I am pushing this branch as is.

Thank you for listening to my TED talk.
I wish you all a wonderful day ahead

kiwibang committed:
Merge pull request #3 from kiwibang/branch-A-NaturalDates

Add BCD-Extension features

Commit:
Add A-BetterGui features

Commit:
Merge branch 'branch-A-BetterGui'

Commit:
Add A-Personality features

Commit:
Merge branch 'branch-A-Personality'

Commit:
Add A-MoreErrorHandling features

Commit:
Add A-MoreTesting features

Commit:
Add A-MoreTesting features

Commit:
Merge branch 'branch-A-MoreTesting'

Commit:
Add A-UserGuide features

kiwibang committed:
Update README.md

Commit:
Add A-Release features

Commit:
Edited gradle.build

Commit:
Add fixes to build.gradle

Commit:
Give credit for build.gradle
